File: 450-delete-node-in-a-bst/delete-node-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
 TreeNode* helper(TreeNode* root){
    if(!root->right && !root->left) return NULL;
   if(!root->left){
    return root->right;
   }else if(!root->right){
    return root->left;
   }else{
    TreeNode* rightguy = root->right;
    TreeNode* ans = root->left;
    TreeNode* rightofLeft = getRight(root->left);
    rightofLeft->right = rightguy;
    return ans;
   }
 }
 TreeNode* getRight(TreeNode* root){
   TreeNode* node = root;
   while(node->right){
    node = node->right;
   }
   return node;
 }
    TreeNode* deleteNode(TreeNode* root, int key) {
        if(!root) return NULL;
        if(root->val==key){
           return helper(root);
        }
        TreeNode* dummy = root;
        while(root){
            if(root->val>key){
          if(root->left && root->left->val==key){
                root->left = helper(root->left);
                break;
            }else{
                root = root->left;
            }
        }else{
            if(root->right && root->right->val==key){
                root->right = helper(root->right);
                break;
            }else{
                root = root->right;
            }
        }
            
        }
        return dummy;
    }
};

File: 450-delete-node-in-a-bst/delete-node-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    TreeNode parent;
    public TreeNode solve(TreeNode root){
        if(root.left==null && root.right==null) return null;
        
        if(root.left==null){
            return root.right;
        }
        if(root.right==null){
            return root.left;
        }
        TreeNode temp = getRight(root.left);
        temp.right = root.right;
        return root.left;
    }
    public TreeNode getRight(TreeNode root){
        TreeNode temp = root;
        while(temp.right!=null){
            temp = temp.right;
        }
        return temp;
    }
    public TreeNode deleteNode(TreeNode root, int key) {
        TreeNode temp = root;
        parent = root;
        if(root==null) return null;
        if(root.val==key){
            return solve(root);
        }
        while(temp!=null){
            if(temp.val>key){
                parent = temp;
                temp = temp.left;
            }else if(temp.val<key){
                parent = temp;
                temp = temp.right;
            }else{
             if(parent.left!=null && parent.left.val==temp.val){
                    parent.left = solve(temp);
                }else if(parent.right!=null && parent.right.val==temp.val){
                    parent.right = solve(temp);
                }
                break;
            }
        }
        return root;
    }
}